`air language-server` doesn't work with helix editor

I've tried to make `air language-server` work under helix editor, but without success so far. I'm not sure what am I messing up.

I've configured `air` similarly to `ruff`, here's my `language.toml`:
```
[language-server.air]
command = "air"
args = ["language-server"]

[[language]]
name = "r"
language-servers = ["air"]
```

When I open a .R file in helix, the language server immediately crashes. I've also ran helix with the `-vvv` flag for the most detailed logging and this is what I got:
```
2025-02-20T16:03:10.961 helix_lsp [INFO] stopping existing clients for 'air'
2025-02-20T16:03:10.961 helix_lsp::file_event [DEBUG] Removing LSP client: 1v1
2025-02-20T16:03:10.961 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","method":"shutdown","id":1}
2025-02-20T16:03:10.962 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:10.962099734+01:00  INFO  Incoming: Shutdown\n"
2025-02-20T16:03:10.962 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/tower_lsp.rs:161\n"
2025-02-20T16:03:10.962 helix_lsp::transport [ERROR] air err <- "\n"
2025-02-20T16:03:10.962 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","method":"textDocument/didClose","params":{"textDocument":{"uri":"file:///home/tp/test.R"}}}
2025-02-20T16:03:10.962 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"general":{"positionEncodings":["utf-8","utf-32","utf-16"]},"textDocument":{"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}},"dataSupport":true,"disabledSupport":true,"isPreferredSupport":true,"resolveSupport":{"properties":["edit","command"]}},"completion":{"completionItem":{"deprecatedSupport":true,"insertReplaceSupport":true,"resolveSupport":{"properties":["documentation","detail","additionalTextEdits"]},"snippetSupport":true,"tagSupport":{"valueSet":[1]}},"completionItemKind":{}},"formatting":{"dynamicRegistration":false},"hover":{"contentFormat":["markdown"]},"inlayHint":{"dynamicRegistration":false},"publishDiagnostics":{"tagSupport":{"valueSet":[1,2]},"versionSupport":true},"rename":{"dynamicRegistration":false,"honorsChangeAnnotations":false,"prepareSupport":true},"signatureHelp":{"signatureInformation":{"activeParameterSupport":true,"documentationFormat":["markdown"],"parameterInformation":{"labelOffsetSupport":true}}}},"window":{"workDoneProgress":true},"workspace":{"applyEdit":true,"configuration":true,"didChangeConfiguration":{"dynamicRegistration":false},"didChangeWatchedFiles":{"dynamicRegistration":true,"relativePatternSupport":false},"executeCommand":{"dynamicRegistration":false},"fileOperations":{"didRename":true,"willRename":true},"inlayHint":{"refreshSupport":false},"symbol":{"dynamicRegistration":false},"workspaceEdit":{"documentChanges":true,"failureHandling":"abort","normalizesLineEndings":false,"resourceOperations":["create","rename","delete"]},"workspaceFolders":true}},"clientInfo":{"name":"helix","version":"25.01.1 (6304e7b2)"},"processId":578064,"rootPath":"/home/tp","rootUri":null,"workspaceFolders":[]},"id":0}
2025-02-20T16:03:10.962 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:10.962288716+01:00  INFO  Outgoing: Shutdown\n"
2025-02-20T16:03:10.962 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/tower_lsp.rs:175\n"
2025-02-20T16:03:10.962 helix_lsp::transport [ERROR] air err <- "\n"
2025-02-20T16:03:10.962 helix_lsp::transport [INFO] air <- {"jsonrpc":"2.0","result":null,"id":1}
2025-02-20T16:03:10.962 helix_lsp::transport [INFO] air <- null
2025-02-20T16:03:10.962 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","method":"exit"}
2025-02-20T16:03:10.962 helix_lsp::transport [ERROR] air err: <- StreamClosed
2025-02-20T16:03:10.962 helix_lsp [DEBUG] client was already removed
2025-02-20T16:03:10.963 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:10.963934563+01:00  INFO  Updating log level:\n"
2025-02-20T16:03:10.963 helix_lsp::transport [ERROR] air err <- "Log level: Info\n"
2025-02-20T16:03:10.963 helix_lsp::transport [ERROR] air err <- "Dependency log levels: None\n"
2025-02-20T16:03:10.963 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/logging.rs:263\n"
2025-02-20T16:03:10.963 helix_lsp::transport [ERROR] air err <- "\n"
2025-02-20T16:03:10.963 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:10.963977676+01:00  INFO  Outgoing: Initialize\n"
2025-02-20T16:03:10.963 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/tower_lsp.rs:175\n"
2025-02-20T16:03:10.963 helix_lsp::transport [ERROR] air err <- "\n"
2025-02-20T16:03:10.964 helix_lsp::transport [INFO] air <- {"jsonrpc":"2.0","result":{"capabilities":{"positionEncoding":"utf-8","textDocumentSync":2,"documentFormattingProvider":true,"documentRangeFormattingProvider":true,"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true}}},"serverInfo":{"name":"Air Language Server","version":"0.0.0"}},"id":0}
2025-02-20T16:03:10.964 helix_lsp::transport [INFO] air <- {"capabilities":{"documentFormattingProvider":true,"documentRangeFormattingProvider":true,"positionEncoding":"utf-8","textDocumentSync":2,"workspace":{"workspaceFolders":{"changeNotifications":true,"supported":true}}},"serverInfo":{"name":"Air Language Server","version":"0.0.0"}}
2025-02-20T16:03:10.964 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","method":"initialized","params":{}}
2025-02-20T16:03:10.964 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"r","text":"library(ggplot2)\n\na <- 1\nb <- 2\n\na + b\n","uri":"file:///home/tp/test.R","version":0}}}
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:10.964103359+01:00  INFO  Incoming: Initialized\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/tower_lsp.rs:184\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:10.964148736+01:00  INFO  Incoming: DidOpenTextDocument\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/tower_lsp.rs:184\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "\n"
2025-02-20T16:03:10.964 helix_lsp::transport [INFO] air <- {"jsonrpc":"2.0","method":"client/registerCapability","params":{"registrations":[{"id":"air-toml-watcher","method":"workspace/didChangeWatchedFiles","registerOptions":{"watchers":[{"globPattern":"**/air.toml"},{"globPattern":"**/.air.toml"}]}}]},"id":0}
2025-02-20T16:03:10.964 helix_lsp::file_event [DEBUG] Registering didChangeWatchedFiles for client '1v3' with id 'air-toml-watcher'
2025-02-20T16:03:10.964 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","result":null,"id":0}
2025-02-20T16:03:10.964 helix_lsp::transport [INFO] air <- {"jsonrpc":"2.0","method":"workspace/configuration","params":{"items":[{"section":"positron.r.diagnostics.enable"},{"scopeUri":"file:///home/tp/test.R","section":"editor.insertSpaces"},{"scopeUri":"file:///home/tp/test.R","section":"editor.indentSize"},{"scopeUri":"file:///home/tp/test.R","section":"editor.tabSize"},{"section":"air.logLevel"},{"section":"air.dependencyLogLevels"},{"section":"air.syncFileSettingsWithClient"}]},"id":1}
2025-02-20T16:03:10.964 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","result":[null,null,null,null,null,null,null],"id":1}
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:10.964445157+01:00 ERROR  Failure while handling event:\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "invalid type: null, expected a boolean\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/main_loop.rs:288\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "\n"
2025-02-20T16:03:12.919 helix_lsp::file_event [DEBUG] Removing LSP client: 1v3
2025-02-20T16:03:12.919 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","method":"shutdown","id":1}
2025-02-20T16:03:12.919 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:12.919423677+01:00  INFO  Incoming: Shutdown\n"
2025-02-20T16:03:12.919 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/tower_lsp.rs:161\n"
2025-02-20T16:03:12.919 helix_lsp::transport [ERROR] air err <- "\n"
2025-02-20T16:03:12.919 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:12.919541646+01:00  INFO  Outgoing: Shutdown\n"
2025-02-20T16:03:12.919 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/tower_lsp.rs:175\n"
2025-02-20T16:03:12.919 helix_lsp::transport [ERROR] air err <- "\n"
2025-02-20T16:03:12.919 helix_lsp::transport [INFO] air <- {"jsonrpc":"2.0","result":null,"id":1}
2025-02-20T16:03:12.919 helix_lsp::transport [INFO] air <- null
2025-02-20T16:03:12.919 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","method":"exit"}
2025-02-20T16:03:12.920 helix_lsp::transport [ERROR] err: <- Other(failed to send a message to server
2025-02-20T16:03:12.920 helix_lsp::transport [ERROR] air err: <- StreamClosed
```

I guess this is the point where things go awry:
```
2025-02-20T16:03:10.964 helix_lsp::transport [INFO] air -> {"jsonrpc":"2.0","result":[null,null,null,null,null,null,null],"id":1}
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "  2025-02-20T16:03:10.964445157+01:00 ERROR  Failure while handling event:\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "invalid type: null, expected a boolean\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "    at crates/lsp/src/main_loop.rs:288\n"
2025-02-20T16:03:10.964 helix_lsp::transport [ERROR] air err <- "\n"
```

Unfortunately, this is as far as I got. Maybe this is not related to `air` at all, and it's a problem with `helix`'s lsp. 
Let me know if I should provide more logs, etc.

P.S. I really appreciate your initiative to make a new, more modern lsp for R, 1000 thanks for making it!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,8 @@
 
 # Development version
 
+- Language server configuration variables are now fully optional, avoiding issues in editors like Zed or Helix (#246).
+
 
 # 0.4.0
 

--- a/crates/lsp/src/settings_vsc.rs
+++ b/crates/lsp/src/settings_vsc.rs
@@ -27,9 +27,9 @@ pub(crate) struct VscGlobalSettings {
 #[derive(Clone, Debug, FieldNamesAsArray, serde::Serialize, serde::Deserialize)]
 pub(crate) struct VscDocumentSettings {
     // DEV NOTE: Update `section_from_key()` method after adding a field
-    pub insert_spaces: bool,
-    pub indent_size: VscIndentSize,
-    pub tab_size: usize,
+    pub insert_spaces: Option<bool>,
+    pub indent_size: Option<VscIndentSize>,
+    pub tab_size: Option<usize>,
 }
 
 #[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
@@ -42,7 +42,7 @@ pub(crate) enum VscIndentSize {
 #[derive(Clone, Debug, FieldNamesAsArray, serde::Serialize, serde::Deserialize)]
 pub(crate) struct VscDiagnosticsSettings {
     // DEV NOTE: Update `section_from_key()` method after adding a field
-    pub enable: bool,
+    pub enable: Option<bool>,
 }
 
 impl From<VscGlobalSettings> for GlobalSettings {
@@ -70,12 +70,20 @@ impl VscDocumentSettings {
 /// representation. Currently one-to-one.
 impl From<VscDocumentSettings> for DocumentSettings {
     fn from(value: VscDocumentSettings) -> Self {
-        let indent_style = indent_style_from_vsc(value.insert_spaces);
-        let indent_width = indent_width_from_vsc(&value);
+        // Conversion is all or nothing to avoid sending, say, the `indent_size` without
+        // also sending the `tab_size`
+        let (indent_style, indent_width) =
+            match (value.insert_spaces, value.indent_size, value.tab_size) {
+                (Some(insert_spaces), Some(indent_size), Some(tab_size)) => (
+                    Some(indent_style_from_vsc(insert_spaces)),
+                    Some(indent_width_from_vsc(indent_size, tab_size)),
+                ),
+                _ => (None, None),
+            };
 
         Self {
-            indent_style: Some(indent_style),
-            indent_width: Some(indent_width),
+            indent_style,
+            indent_width,
             line_width: None, // We don't currently watch this setting
         }
     }
@@ -101,15 +109,18 @@ impl VscGlobalSettings {
     }
 }
 
-pub(crate) fn indent_width_from_vsc(settings: &VscDocumentSettings) -> settings::IndentWidth {
-    let indent_width = match settings.indent_size {
+pub(crate) fn indent_width_from_vsc(
+    indent_size: VscIndentSize,
+    tab_size: usize,
+) -> settings::IndentWidth {
+    let indent_width = match indent_size {
         VscIndentSize::Size(ref size) => *size,
         VscIndentSize::Alias(ref var) => {
             if var != "tabSize" {
                 tracing::warn!("Unknown indent alias {var}, using default");
                 return settings::IndentWidth::default();
             }
-            settings.tab_size
+            tab_size
         }
     };
 

